Paver: Ability to run only servers or only tests

benpatterson opened the pull request:
This would allow a user to set up and run servers, with an open prompt for killing them.

Likewise a user could open a different terminal session and run tests only.

benpatterson commented:
@mulby 

Here a user would open one ssh session/terminal session and run `paver test_bokchoy --serversonly` (or, to skip collectstatic, `paver test_bokchoy --serversonly --fasttest`). That will leave a session waiting for an enter character.

Meanwhile, in another ssh/terminal session, a user could run `paver test_bokchoy -t <test spec> --testsonly`

Thoughts?

benpatterson commented:
test this please

benpatterson commented:
@jzoldak @clytwynec. Thoughts?

benpatterson commented:
Also reference #7912

mulby commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -94,19 +119,11 @@ def __enter__(self):
+        """
+        Stops running process and waits for user input before continuing.
+        """
+        sh('read -p "Press enter to stop servers." PROMPTVAR')
```
nit: we could probably just use the python native input reading capability here

jzoldak replied:
I'd prefer if you had to ctrl-C in order to stop the server, rather than any input. I think otherwise it is too easy to accidentally hit a key in the wrong terminal window.

jzoldak replied:
Oh, also people have had trouble in the past messing up their local devstack's .noseid files by hitting ^C multiple times quickly to kill running bok-choy tests. Although that falls into some "user error" gray area, we should try to make sure that issue doesn't become *more* prevalent. I'm not sure exactly how, but maybe by displaying text that instructs them to hit ^C once only?

benpatterson replied:
Taking both your and Jay's suggestions, I've changed this in the last commit to a while loop.

mulby commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -58,6 +60,29 @@ def __enter__(self):
+        if self.serversonly:
+            self.interruptable_prompt()
+
+    def __exit__(self, exc_type, exc_value, traceback):
```
Do we want to run all of this when you specify --testsonly? I feel like it should do "no setup" and "no teardown"...

benpatterson replied:
I left this in here since we'd want to know if the servers got killed (or the user forgot to start them). Otherwise finding out would likely happen with some confusing browser exceptions. That said, I've changed the wording so this waiting period is less confusing.

mulby commented:
This would work for me :+1: 

Thanks!

jzoldak commented:
Can you please include in this PR user documentation for the new functionality?

benpatterson commented:
@jzoldak @mulby @clytwynec 

I made improvements based on your comments. Care to take another look?

Since I changed the commit message, I also squashed. Sorry if that makes review a bit trickier! The main focii are the infinite loop for the servers' running, and doc.

jzoldak commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -23,7 +25,9 @@ class BokChoyTestSuite(TestSuite):
       report_dir - directory for reports (e.g., coverage) related to test execution
       xunit_report - directory for xunit-style output (xml)
-      fasttest - when set, skip various set-up tasks (e.g., DB migrations)
+      fasttest - when set, skip various set-up tasks (e.g., collectstatic)
```
collectstatic and DB migrations? or just collectstatic?

benpatterson replied:
Just collectstatic. (And compiling static assets, if I remember right.) DB migrations definitely DO run when using fasttest. That, in part, is because we blow away the database on each (traditional) run.

jzoldak commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -55,9 +61,32 @@ def __enter__(self):
+        if not self.testsonly:
+            self.prepare_bokchoy_run()
+
+        msg = colorize('green', "Confirming servers have started...")
```
The msg would make more sense to me in both contexts if it instead said "Confirming servers are running..."

benpatterson replied:
Done

jzoldak commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -116,19 +145,25 @@ def cmd(self):
-            "--xunit-file={}".format(self.xunit_report),
-            "--verbosity={}".format(self.verbosity),
-        ]
+        else:
```
This doesn't need to be in an else block because the if branch would have already returned.

benpatterson replied:
Good point. This was an artifact of an earlier approach. Fixed.

jzoldak commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -91,22 +120,22 @@ def __enter__(self):
-        bokchoy_utils.clear_mongo()
 
     @property
     def cmd(self):
```
Can you please add a docstring to this?

benpatterson replied:
Done

jzoldak commented:
@Kelketek @mattdrayer @muhammad-ammar  You guys might be interested in this one - it's an efficiency improvement for the process of writing/debugging bok-choy tests that @mulby suggested.

jzoldak commented:
:+1:

mattdrayer commented:
Wow, great!  Also I noticed the comment about the corrupted .noseids file
due to hitting Ctrl-C too many times -- I'm an offender of this and it
actually hosed me yesterday for a couple hours but now am aware of what
causes the problem.

On Wed, May 6, 2015 at 10:22 AM, Jesse Zoldak <notifications@github.com>
wrote:

> [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7916#issuecomment-99485907>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

mattdrayer commented:
@ziafazal @asadiqbal08 -- FYI on this helpful PR

benpatterson commented:
Ah ok, hearing another vote for that...and considering the .noseids problem is particularly nasty, I decided to put in a message about it. I think there's very little risk about the 'serversonly' process causing problems with multiple Ctrl-C's. But, we can just beat the drum a bit so people are reminded about it in general.

jzoldak commented:
@benpatterson I think in this case it would happen if people hit ^C multiple times on the testsonly side.
In some testing of this PR today I totally understand hitting it multiple times, as there is a big perceived lag on that side between when you first hit ^C and the database cleanup stuff finishes. Although thinking it through, my guess is that by then then nose command has finished, so if you hit ^C again at that time it will only mess up / not do the database flush, and will not mess up the .noseid files. FWIW.

muhammad-ammar commented:
@benpatterson Awesome Work.
